src/common/openshiftHelpers.js

```diff
@@ -0,0 +1,99 @@
+import { create, list } from '../services/openshiftServices';
+
+/**
+ * Construct a projects namespace from a given username.
+ * Note that the namespace name might contain the full username as it is sanitized first.
+ * @param {string} username The username to create the namespace name from.
+ * @param {string} suffix A suffix to append to the end of the users namespace.
+ */
+const buildValidProjectNamespaceName = (username, suffix) => `${cleanUsername(username)}-${suffix}`;
+
+/**
+ * Get a sanitized version of a username, so it can be used to name OpenShift.
+ * @param {string} username The username to sanitize.
+ */
+const cleanUsername = username => username.replace(/@/g, '-').replace(/\./g, '-');
+
+/**
+ * Helper function for finding a single OpenShift Resource.
+ * @param {Object} resourceDef The definition of the OpenShift Resource.
+ * @param {Object} resToFind The OpenShift Resource itself. By default this needs only to contain a name.
+ * @param {Function} compareFn A custom function for comparing resources, determines if a resource is found.
+ */
+const findOpenshiftResource = (
+  resourceDef,
+  resource,
+  compareFn = resObj => resObj.metadata.name === resource.metadata.name
+) =>
+  list(resourceDef)
+    .then(listResponse => (listResponse && listResponse.items ? listResponse.items : []))
+    .then(resourceObjs => resourceObjs.find(resObj => compareFn(resObj)));
+
+/**
+ * Helper function for creating an OpenShift Resource if it doesn't exist already.
+ * @param {Object} resourceDef The definition of the OpenShift Resource.
+ * @param {Object} resToFind The OpenShift Resource itself. By default this needs only to contain a name.
+ * @param {Function} compareFn A custom function for comparing resources, determines if a resource is found.
+ */
+const findOrCreateOpenshiftResource = (resourceDef, resource, compareFn) =>
+  findOpenshiftResource(resourceDef, resource, compareFn).then(foundResource => {
+    if (!foundResource) {
+      return create(resourceDef, resource);
+    }
+    return Promise.resolve(foundResource);
+  });
+
+const namespaceRequestDef = {
+  name: 'projectrequests',
+  version: 'v1',
+  group: 'project.openshift.io'
+};
+const namespaceRequestResource = name => ({
+  kind: 'ProjectRequest',
+  metadata: {
```

Comment:
do we need projectRequests? we can use kind project to create new namespace - https://docs.openshift.com/container-platform/3.10/rest_api/apis-project.openshift.io/v1.Project.html#Post-apis-project.openshift.io-v1-projects
check #56 openshiftKinds.js,  middlewareServices.js walkthroughServices.js for examples

Reply:
@witmicko Have you tried running as a non-admin user? From what I remember non-admin users use ProjectRequests when creating a namespace

Reply:
I ll that that now @aidenkeating